### DISCUSSION TITLE
How to control icons in OUTLINE viewlet?

### DISCUSSION THREAD

**Discussion by Serhioromano** [eclipse-langium/langium#796] [Q&A] [answered]
I am starting to make progress. Thank you for your help. Outline is generated now like this.

![2022-11-29_18-06-27](https://user-images.githubusercontent.com/650741/204525216-3aba57f7-3dad-4bb9-ab84-8029e22bd0f7.png)

How can I change the icons? Whose are different things like classes, functions, interfaces, ....

Also last elements there are generated from 

```
FUNCTION_BLOCK anotherblock
    VAR CONSTANT
        Variable: BYTE; (* Comment *)
        Variable2: MyBlock9; (* Comment *)
    END_VAR
    VAR
        Variable: BOOL; (* Comment *)
    END_VAR

    METHOD Name : BOOL
        VAR
            Variable: BOOL; (* Comment *)
        END_VAR
    END_METHOD
END_FUNCTION_BLOCK
```

But I want tree to be like this

```
+ anotherblock (Block)
| + Var Constant
| | - Variable
| + Var
| | - Variable
| + Name (Method)
| | + Var
| | | - Variable
```

**Comment by msujew** [comment, marked as answer]
Hey @Serhioromano,

you will need to override the [`DefaultDocumentSymbolProvider`](https://github.com/langium/langium/blob/main/packages/langium/src/lsp/document-symbol-provider.ts) class and bind it in your module (similar to [here](https://github.com/langium/langium/blob/5ab46cf82543c75c26750a38cca4b00323fd7b2a/packages/langium/src/grammar/langium-grammar-module.ts#L36)). Especially overriding the [`getSymbolKind`](https://github.com/langium/langium/blob/5ab46cf82543c75c26750a38cca4b00323fd7b2a/packages/langium/src/lsp/document-symbol-provider.ts#L71-L73) method is reasonable for your case.

**Reply by snarkipus** [reply]
This is an example I put together while working through/learning this same feature.

[Custom Document Symbol Provider](https://github.com/snarkipus/small-java/blob/main/src/language-server/small-java-document-symbol-provider.ts)
